subsys/fs/littlefs_fs.c

```diff
@@ -595,8 +613,42 @@ static int littlefs_mount(struct fs_mount_t *mountp)
 	dev = flash_area_get_device(fs->area);
 	if (dev == NULL) {
 		LOG_ERR("can't get flash device: %s", log_strdup(fs->area->fa_dev_name));
-		ret = -ENODEV;
-		goto out;
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int littlefs_mount(struct fs_mount_t *mountp)
+{
+	int ret = 0;
+	struct fs_littlefs *fs = mountp->fs_data;
+
+	LOG_INF("LittleFS version %u.%u, disk version %u.%u",
+		LFS_VERSION_MAJOR, LFS_VERSION_MINOR,
+		LFS_DISK_VERSION_MAJOR, LFS_DISK_VERSION_MINOR);
+
+	if (littlefs_on_blkdev(mountp)) {
```

Comment:
This would be already fixed in the `littlefs_on_blkdev` function which for `CONFIG_FS_LITTLEFS_BLK_DEV` not defined will just return false.

Reply:
> 

No it is not. The `IS_ENABLED` thing checks if KConfig option is enabled or not and basically evaluates to `true` or `false`.
If you do the `if (IS_ENABLED(sth) && some_call(...))`, what you get is either:
```
if (0 && some_call())
```
or
```
if (1 && some_call())
```

You can can probably now predict what will compiler do with the code, seeing literal being "&&" with the rest of expression: it will either remove entire block, if literal is 0, or leave non-literal part of expression only, when the literal is 1.

This is "#if/#ifdef" being passed to compiler.

That is why I have also proposed this https://github.com/zephyrproject-rtos/zephyr/pull/40704#discussion_r758398897, because instead of
```
if (0 && some_call()) {
}
```
you will just have
```
if (0 && 0) {
}
```
where `some_call`, in the former, requires some definition for the linker, while the later does not.